doc/private/man/man3/SubmitTransferCli.3
.TH "SubmitTransferCli" 3 "Wed Feb 8 2012" "Version 0.0.0" "EMI File Transfer Service (FTS 3)" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SubmitTransferCli \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBCliBase\fP.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBtransfer__TransferParams\fP * \fBgetParams\fP (soap *soap)"
.br
.ti -1c
.RI "bool \fBperformChecks\fP ()"
.br
.ti -1c
.RI "virtual void \fBinitCli\fP (int ac, char *av[])"
.br
.ti -1c
.RI "string \fBgetPassword\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 15 of file SubmitTransferCli.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for EMI File Transfer Service (FTS 3) from the source code.


fts2-prototype/RELEASE-NOTES
*** Release glite-data-transfer-cli_R_4_0_2 2011-10-14
	#109 	glite-transfer-submit man page issues https://svnweb.cern.ch/trac/glitefts/ticket/109

*** Release glite-data-transfer-cli_R_4_0_0 2011-03-07

    o glite-transfer-channel-audit fails with Segmentation Fault  (bug#68524)
    o Tool to modified automatically the state of the channels given an SE  (bug#61780)    


*** Release glite-data-transfer-cli_R_3_7_5_1 2010-11-29

    o changed util.c to output from stdout to stderr  (bug#75722)

*** Release glite-data-transfer-cli_R_3_7_4_1 2010-01-22

    o added glite-transfer-se-set (bug#61780)

*** Release glite-data-transfer-cli_R_3_7_3_1 2009-12-17

	o Fix for bug #52416: Bad error treatment for ExistsException on the client	side
	o Minimal Python API with test.

*** Release glite-data-transfer-cli_R_3_7_2_1 2009-08-28

    o bug #41584: [delegation] IPv6 bug: non compliant name resolving 
      function in source code (gethostbyname_r)
      This bug also applies to this component. 

*** Release glite-data-transfer-cli_R_3_7_1_1 2009-08-05

	o Fixing a regression: overwrite flag (-o) should not require an argument
    o Updated the test suite to the latest FTS service.

*** Release glite-data-transfer-cli_R_3_7_0_1 2009-07-03

	o v3.7.0 of the interface with checksum support.

	  glite-transfer-submit can take the -K option for checksum comparison
	  and can pass on optional checksums (algorithm:hexadecimal-numbers)
	  for each source-destination SURL pair given by the command line or
	  in a job description file.

*** Release glite-data-transfer-cli_R_3_6_3_1 2009-03-30
	o fixed bug #48684: transfer-cli should obsolete transter-api-c and srm-cli packages

*** Release glite-data-transfer-cli_R_3_6_2_1 2009-03-22
    
	o Fixed Savannah bug #35997:  FTS cli: if an invalid value is passed
	  with -s option, the default is used
    o Removed obsolete commands: glite-transfer-placement-submit
      and glite-transfer-discovery

*** Release glite-data-transfer-cli_R_3_6_0_1 2009-01-09
    o fixed bug 45167
    o fixed bug 28212
    o fixed bug #39819
	o added site group management tools
	o Fixed the seg fault for --tx-log-level and --srmcopy-direction options
	o Fixed the option --urlcopy-tx-to-per-mb that was executing --urlcopy-no-prog-to
	o Fixed libraries build order (avoid relinking and appending U to .so file names)

*** Release glite-data-transfer-cli_R_3_5_2_1 2008-12-04
	o Fixed bug 41682

*** Release glite-data-transfer-cli_R_3_5_1_1 2008-12-03
	o slc5 fixes

*** Release glite-data-transfer-cli_R_3_5_0_1 2008-11-27
    o Merged transfer-api-c into transfer-cli.
    o updated to schema version 3.5.0

*** Release glite-data-transfer-cli_R_3_4_5_1 2008-08-11
    o fixed bug #40064 
	o fixed bug #39538
    o refactored options parsing 
    o modified help messages

*** Release glite-data-transfer-cli_R_3_4_4_3 2008-06-17
    o Added dependency from gsoap

*** Release glite-data-transfer-cli_R_3_4_4_2 2008-05-08
	o Fix bug #33406

*** Release glite-data-transfer-cli_R_3_4_3_2 2008-03-20
	o changes to command line help in submit - clearer token handling

*** Release glite-data-transfer-cli_R_3_4_3_1 2008-03-18
	 #30034: FTS: cannot set numerical channel properties to zero
	 #33406: FTS: CLI version string is not reported correctly in verbose option
	 #28524: Missing -o parameter from glite-transfer-submit help
	 Interface changes related to the latest schema changes 3.2.0
	 Number of undocumented bugs fixed related to the marshalling.

*** Release glite-data-transfer-cli_R_3_4_2_2 2007-06-06
    o lib64 changes

*** Release glite-data-transfer-cli_R_3_4_2 2007-06-01

   o Added overwrite flag for SRM 2.2. Proper name for space token description.

*** Release glite-data-transfer-cli_R_3_4_1-1 2007-04-20

   o Fixed SLC4 build error and a few compiler warnings.

*** Release glite-data-transfer-cli_R_3_3_0-4 2007-03-02
   o Compile with gsoap 2.7.6b and 2.7.9b 

*** Release glite-data-transfer-cli_R_3_3_0-1 2006-08-22
   o Add option -x to the glite-transfer-discovery tool in order to display all the 
     FTS service information
   o Add option --all to the glite-transfer-discovery tool in order to return
     all the FTS services that are able to manage a transfer between the given two sites
   o Fix for #19196: FTS: cli should support site names in teh -s option

*** Release glite-data-transfer-cli_R_3_2_3_4 2006-07-06

   o Fix for #17975 - service info interface version bad
   o Fix for #16546 - bad error on addChannel
   o Make interesting states (default serach) only Active, Submitted and Pending
   o Better date format on stats
   o Fix for #16581 - unclear manpage for addManager

*** Release glite-data-transfer-cli_R_3_2_3-3 2006-06-21

	o Merge changes from 3.2.2-3

*** Release glite-data-transfer-cli_R_3_2_2-3 2006-06-21
    Fixed Bugs:
     o #17608: FTS: glite-transfer-stats-vo segfaults

*** Release glite-data-transfer-cli_R_3_2_3-2 2006-06-14

	o Merge changes from 3.2.2-2

*** Release glite-data-transfer-cli_R_3_2_2-2 2006-06-14
    Fixed Bugs:
     o #17460: FTS: glite-transfer-discovery display the wrong destination site when the --verbose option is used

*** Release glite-data-transfer-cli_R_3_2_3 2006-05-02

	o Added Doxygen API documentation for the simple transfer API.

*** Release glite-data-transfer-cli_R_3_2_2 2006-04-28
    Add command line options -T and -S to glite-transfer-discovery
    Fixed Bugs:
     o #16424 FTS: BDII service discovery should work 
     o #16215: fix man page of glite-transfer-channel-list 

*** Release glite-data-transfer-cli_R_3_2_1 2006-03-23

    o CGSI_gSOAP library location has changed from ${repository}/${ext.cgsi-gsoap.subdir} to ${with.cgsi-gsoap.prefix}

*** Release glite-data-transfer-cli_R_3_2_0 2006-02-22
    Add glite-transfer-discovery tool for returning the endpoint of an
       FTS service that manage a channel between two sites
    Fixed bugs:
     o #14765 FTS: Logical name is not displayed for placement jobs


src/CLI/SubmitTransferCli.cpp
/*
 * SubmitTransferCli.cpp
 *
 *  Created on: Feb 2, 2012
 *      Author: simonm
 */

#include "SubmitTransferCli.h"
#include "ftsH.h"
#include "CliManager.h"

#include <iostream>
#include <fstream>

SubmitTransferCli::SubmitTransferCli() {

	specific.add_options()
			("blocking,b", "Blocking mode, wait until the operation completes.")
			("file,f", value<string>(&cfg_file), "Name of a configuration file.")
			("gparam,g", value<string>(), "Gridftp parameters.")
			("interval,i", value<int>(), "Interval between two poll operations in blocking mode.")
			("myproxysrv,m", value<string>(), "MyProxy server to use.")
			("password,p", value<string>(), "MyProxy password to send with the job")
			("id,I", value<string>(), "Delegation with ID as the delegation identifier.")
			("expire,e", value<long>(), "Expiration time of the delegation in minutes.")
			("overwrite,o", "Overwrite files.")
			("dest-token,t", value<string>(),  "The destination space token or its description (for SRM 2.2 transfers).")
			("source-token,S", value<string>(), "he source space token or its description (for SRM 2.2 transfers).")
			("compare-checksum,K", "Compare checksums between source and destination.")
			("copy-pin-lifetime", value<int>(), "pin lifetime of the copy of the file (seconds)")
			("lan-connection", "use LAN as ConnectionType (default = WAN)")
			("fail-nearline", "fail the transfer if the file is nearline")
			;

	cli_options.add(specific);
}

SubmitTransferCli::~SubmitTransferCli() {

}

void SubmitTransferCli::initCli(int ac, char* av[]) {
	CliBase::initCli(ac, av);

	if (vm.count("id")) {
		delegate = true;
	}

	CliManager* manager = CliManager::getInstance();

    // decide on delegation depending on server
	if (manager->isDelegationSupported()) {
		if (!vm.count("password")) {
	        cout << "# Server supports delegation. Delegation will be used by default." << endl;
	        delegate = true;
		} else {
	        cout << "# Server supports delegation, however a MyProxy pass phrase was given: will use MyProxy legacy mode." << endl;
	        delegate = false;
		}
	}

    if (vm.count("password")) {
    	password = vm["password"].as<string>();
    } else {
    	if (!delegate) {
    		password = CliManager::getInstance()->getPassword();
    	}
    }

    /*if (!cfg_file.empty()) {
    	cfg_desc.add(cli_options);

        ifstream ifs(cfg_file.c_str());
        if (!ifs) {
            cout << "Can not open configuration file: " << cfg_file << endl;
            return 0;
        } else {
            store(parse_config_file(ifs, cfg_desc), vm);
            notify(vm);
        }
    }*/
}

bool SubmitTransferCli::performChecks() {

	CliManager* manager = CliManager::getInstance();

	// the job is not specified
	if ((getSource().empty() || getDestination().empty()) && !vm.count("file")) {
		cout << "No transfer job is specified." << endl;
		return false;
	}

	// the job is specified twice using cli and in the config file
	if ((!getSource().empty() || !getDestination().empty()) && vm.count("file")) {
		cout << "You may not specify a transfer on the command line if the -f option is used." << endl;
		return false;
	}

	if (vm.count("compare-checksum") && !manager->isChecksumSupported()) {
		cout << "You have specified an optional checksum, but it is not supported by the server." << endl;
		return false;
	}

    // in the case of the -I option, delegation is required
    if (delegate && !manager->isDelegationSupported()) {
        cout <<"The server you are contacting does not support credential delegation (it is running interface version ";
        cout << /*srv_iface_ver <<*/ "). Please use the MyProxy password legacy method for this server." << endl;
        return false;
    }

    // in the case of the -K option, checksum is required
    if (vm.count("compare-checksum") && !manager->isChecksumSupported()) {
        cout <<"The server you are contacting does not support checksum checking (it is running interface version ";
        cout << /*srv_iface_ve <<*/ "). Please omit checksums for this server." << endl;
        return false;
    }

	return true;
}

transfer__TransferParams* SubmitTransferCli::getParams(soap* soap) {

	CliManager* manager = CliManager::getInstance();
	transfer__TransferParams *jobParams = soap_new_transfer__TransferParams(soap, -1);

	if (vm.count("compare-checksum")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_CHECKSUM_METHOD);
		jobParams->values.push_back("compare");
	}

	if (vm.count("overwrite")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_OVERWRITEFLAG);
		jobParams->values.push_back("Y");
	}

	if (vm.count("lan-connection")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_LAN_CONNECTION);
		jobParams->values.push_back("Y");
	}

	if (vm.count("fail-nearline")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_FAIL_NEARLINE);
		jobParams->values.push_back("Y");
	}

	if (vm.count("gparam")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_GRIDFTP);
		jobParams->values.push_back(vm["gparam"].as<string>());
	}

	if (vm.count("myproxysrv")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_MYPROXY);
		jobParams->values.push_back(vm["myproxysrv"].as<string>());
	}

	if (vm.count("id")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_DELEGATIONID);
		jobParams->values.push_back(vm["id"].as<string>());
	}

	if (vm.count("dest-token")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_SPACETOKEN);
		jobParams->values.push_back(vm["dest-token"].as<string>());
	}

	if (vm.count("source-token")) {
		jobParams->keys.push_back(manager->FTS3_PARAM_SPACETOKEN_SOURCE);
		jobParams->values.push_back(vm["source-token"].as<string>());
	}

	if (vm.count("copy-pin-lifetime")) {
		string value = boost::lexical_cast<string>(vm["copy-pin-lifetime"].as<int>());
		jobParams->keys.push_back(manager->FTS3_PARAM_COPY_PIN_LIFETIME);
		jobParams->values.push_back(value);
	}

	return jobParams;
}

string SubmitTransferCli::getPassword() {
	return password;
}
